El/pkce gem pr

RealNobody opened the pull request:
### Summary

Doorkeeper 5.0 had several changes, including PKCE support which are required for the ORM to work properly.

I did a diff of doorkeeper 4.2.6 and 5.0 to determine what changes we would need:
https://github.com/doorkeeper-gem/doorkeeper/compare/v4.2.6...master?expand=1

I concentrated on the mixin changes which represent the major changes that this gem should be concerned with.

I considered using the Doorkeeper mixins now that they do not include ActiveRecord, but felt that it was best to do that at a later time.

### Other Information

I was able to get the master branch of doorkeeper tests to run locally, but do not know how to get CI tests to work.

Thanks for contributing to Doorkeeper-MongoDB project!

RealNobody commented:
This is the same PR as https://github.com/doorkeeper-gem/doorkeeper-mongodb/pull/33, just a different branch for internal reasons in my office.

nbulaj commented:
Hi @RealNobody . Could you please rebase your branch?

Tomorrow I will release mongodb extension compatible with Doorkeeper 4.4 and then we can continue to work on 5.0 compatible version

nbulaj commented:
4.2.0 released

nbulaj commented:
`master` updated to Doorkeeper 5.0, rebase please

RealNobody commented:
> rebase please

Rebased.  I see some things I added to this PR were in master, but not most.  I think this is ready for you again.

Thanks, and enjoy!

nbulaj commented on lib/doorkeeper/orm/mongoid5/stale_records_cleaner.rb:
```diff
@@ -0,0 +1,21 @@
+
+module Doorkeeper
+  module Orm
+    module ActiveRecord
```
Maybe `Mongoid5` ? :)

nbulaj commented on lib/doorkeeper/orm/mongoid4/stale_records_cleaner.rb:
```diff
@@ -0,0 +1,21 @@
+
+module Doorkeeper
+  module Orm
+    module ActiveRecord
```
Mongoid4 ?

nbulaj commented on lib/doorkeeper/orm/mongo_mapper/stale_records_cleaner.rb:
```diff
@@ -0,0 +1,21 @@
+
+module Doorkeeper
+  module Orm
+    module ActiveRecord
```
`MongoMapper`

nbulaj commented on lib/doorkeeper/orm/mongoid6/stale_records_cleaner.rb:
```diff
@@ -0,0 +1,21 @@
+
+module Doorkeeper
+  module Orm
+    module ActiveRecord
```
`Mongoid6`

nbulaj commented on lib/doorkeeper/orm/mongoid7/stale_records_cleaner.rb:
```diff
@@ -0,0 +1,21 @@
+
+module Doorkeeper
+  module Orm
+    module ActiveRecord
```
`Mongoid7`

RealNobody replied:
@nbulaj I forgot to comment on the items you're commenting on.

At the time I wrote this, master in https://github.com/doorkeeper-gem/doorkeeper did not support an ORM specific StaleRecordsCleaner, so I wrote it this way to allow tests to pass.

I now see that a change was made to the doorkeeper gem to allow ORM versions of this class, and I will fix.

Thanks.

nbulaj commented on spec/support/orm/mongo_mapper.rb:
```diff
@@ -7,4 +7,20 @@
     Doorkeeper::AccessToken.create_indexes
   end
+
+  # This is a hack to fix a bug in a spec test in Doorkeeper.
```
Already merged to master

RealNobody replied:
Now, it wasn't when I wrote it.  I'll take it out.

nbulaj commented on spec/support/orm/mongoid.rb:
```diff
@@ -7,4 +7,20 @@
     Doorkeeper::AccessToken.create_indexes
   end
+
+  # This is a hack to fix a bug in a spec test in Doorkeeper.
```
Merged to master

nbulaj commented on lib/doorkeeper-mongodb/version.rb:
```diff
@@ -5,9 +5,9 @@ def self.gem_version
-    TINY = 0
+    MAJOR = 5
+    MINOR = 0
+    TINY = 2
```
`TINY` leave as `0`

RealNobody replied:
Will fix

nbulaj commented on lib/doorkeeper-mongodb/mixins/mongoid/base_mixin.rb:
```diff
@@ -9,6 +9,14 @@ def ordered_by(attribute, direction = :asc)
+        def as_json(*args)
+          json_response = super
+
+          json_response["id"] = json_response["_id"]
```
Why do we need this information? :thinking: Original Doorkeepr token mixin doesn't return internal ID column in JSON response

RealNobody replied:
There was a spec in doorkeeper that was specifically testing for the field `id`.  When Mongoid serializes it only serialized a `_id`.  I added this to allow doorkeeper specs to pass.

Other things changed since I wrote this PR, so maybe this has changed too.  If I can get tests to pass without it, I'll take it out.

You asked though, so that is why.

nbulaj commented:
Thanks @RealNobody ! I'll do some other required fixes when I will have free time :)

nbulaj commented:
@RealNobody thank you for your work! 👍
